images/virtualization-artifact/pkg/controller/vi/vi_reconciler.go

```diff
@@ -0,0 +1,190 @@
+	handlers []Handler
+	client   client.Client
+}
+
+func NewReconciler(client client.Client, handlers ...Handler) *Reconciler {
+	return &Reconciler{
+		client:   client,
+		handlers: handlers,
+	}
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	vi := service.NewResource(req.NamespacedName, r.client, r.factory, r.statusGetter)
+
+	err := vi.Fetch(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if vi.IsEmpty() {
+		return reconcile.Result{}, nil
+	}
+
+	var requeue bool
+
+	slog.Info("Start reconcile VI", slog.String("namespacedName", req.String()))
+
+	var handlerErrs []error
+
+	for _, h := range r.handlers {
+		slog.Info("Run handler", slog.String("name", h.Name()))
+		var res reconcile.Result
+		res, err = h.Handle(ctx, vi.Changed())
+		if err != nil {
+			slog.Error("Failed to handle vi", "err", err)
+			handlerErrs = append(handlerErrs, err)
+		}
+
+		requeue = requeue || res.Requeue
+	}
+
+	vi.Changed().Status.ObservedGeneration = vi.Changed().Generation
+
+	err = vi.Update(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	err = errors.Join(handlerErrs...)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if requeue {
+		slog.Info("Requeue for VI", slog.String("namespacedName", req.String()))
+		return reconcile.Result{
+			RequeueAfter: 5 * time.Second,
+		}, nil
+	}
+
+	slog.Info("Finished reconcile VI", slog.String("namespacedName", req.String()))
+	return reconcile.Result{}, nil
+}
+
+func (r *Reconciler) SetupController(_ context.Context, mgr manager.Manager, ctr controller.Controller) error {
+	if err := ctr.Watch(
+		source.Kind(mgr.GetCache(), &virtv2.VirtualImage{}),
+		&handler.EnqueueRequestForObject{},
+		predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool { return true },
+			DeleteFunc: func(e event.DeleteEvent) bool { return true },
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				return e.ObjectOld.GetGeneration() != e.ObjectNew.GetGeneration()
+			},
+		},
+	); err != nil {
+		return err
+	}
+
+	if err := ctr.Watch(
+		source.Kind(mgr.GetCache(), &cdiv1.DataVolume{}),
+		handler.EnqueueRequestForOwner(
+			mgr.GetScheme(),
+			mgr.GetRESTMapper(),
+			&virtv2.VirtualImage{},
+			handler.OnlyControllerOwner(),
+		),
+		predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool { return false },
+			DeleteFunc: func(e event.DeleteEvent) bool { return false },
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				oldDV, ok := e.ObjectOld.(*cdiv1.DataVolume)
+				if !ok {
+					return false
+				}
+				newDV, ok := e.ObjectNew.(*cdiv1.DataVolume)
+				if !ok {
+					return false
+				}
+
+				return oldDV.Status.Progress != newDV.Status.Progress
+			},
+		},
+	); err != nil {
+		return fmt.Errorf("error setting watch on DV: %w", err)
+	}
+
+	if err := ctr.Watch(
+		source.Kind(mgr.GetCache(), &corev1.PersistentVolumeClaim{}),
+		handler.EnqueueRequestForOwner(
+			mgr.GetScheme(),
```

Comment:
Watches on DataVolume and PVC are useless now. VirtualImage has only one target: "ContainerRegistry".